types/set.go

```diff
@@ -0,0 +1,276 @@
+	set := Set{
+		ElemType: t.ElemType,
+	}
+	if !in.IsKnown() {
+		set.Unknown = true
+		return set, nil
+	}
+	if in.IsNull() {
+		set.Null = true
+		return set, nil
+	}
+	val := []tftypes.Value{}
+	err := in.As(&val)
+	if err != nil {
+		return nil, err
+	}
+	elems := make([]attr.Value, 0, len(val))
+	for _, elem := range val {
+		av, err := t.ElemType.ValueFromTerraform(ctx, elem)
+		if err != nil {
+			return nil, err
+		}
+		elems = append(elems, av)
+	}
+	set.Elems = elems
+	return set, nil
+}
+
+// Equal returns true if `o` is also a SetType and has the same ElemType.
+func (t SetType) Equal(o attr.Type) bool {
+	if t.ElemType == nil {
+		return false
+	}
+	other, ok := o.(SetType)
+	if !ok {
+		return false
+	}
+	return t.ElemType.Equal(other.ElemType)
+}
+
+// ApplyTerraform5AttributePathStep applies the given AttributePathStep to the
+// set.
+func (t SetType) ApplyTerraform5AttributePathStep(step tftypes.AttributePathStep) (interface{}, error) {
+	if _, ok := step.(tftypes.ElementKeyValue); !ok {
+		return nil, fmt.Errorf("cannot apply step %T to SetType", step)
+	}
+
+	return t.ElemType, nil
+}
+
+// Validate implements type validation. This type requires all elements to be
+// unique.
+func (s SetType) Validate(ctx context.Context, in tftypes.Value, path *tftypes.AttributePath) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	if !in.Type().Is(tftypes.Set{}) {
+		err := fmt.Errorf("expected Set value, received %T with value: %v", in, in)
+		diags.AddAttributeError(
+			path,
+			"Set Type Validation Error",
+			"An unexpected error was encountered trying to validate an attribute value. This is always an error in the provider. Please report the following to the provider developer:\n\n"+err.Error(),
+		)
+		return diags
+	}
+
+	if !in.IsKnown() {
+		return diags
+	}
+
+	var elems []tftypes.Value
+
+	if err := in.As(&elems); err != nil {
+		diags.AddAttributeError(
+			path,
+			"Set Type Validation Error",
+			"An unexpected error was encountered trying to validate an attribute value. This is always an error in the provider. Please report the following to the provider developer:\n\n"+err.Error(),
+		)
+		return diags
+	}
+
+	// Attempting to use map[tftypes.Value]struct{} for duplicate detection yields:
+	//   panic: runtime error: hash of unhashable type tftypes.primitive
+	// Instead, use for loops.
+	for indexOuter, elemOuter := range elems {
+		// Only evaluate fully known values for duplicates.
+		if !elemOuter.IsFullyKnown() {
+			continue
+		}
+
+		for indexInner, elemInner := range elems {
+			if indexInner == indexOuter {
+				continue
+			}
+
+			if !elemInner.Equal(elemOuter) {
+				continue
+			}
+
+			diags.AddAttributeError(
+				path.WithElementKeyValue(elemInner),
+				"Duplicate Set Element",
+				fmt.Sprintf("This attribute contains duplicate values of: %s", elemInner),
+			)
+		}
+	}
+
+	return diags
+}
+
+// Set represents a set of AttributeValues, all of the same type, indicated
+// by ElemType.
+type Set struct {
```

Comment:
Do we want to add helper methods for:

```go
// Add inserts an element into the Set, throwing an error if it's already present
func (s Set) Add(ctx context.Context, el attr.Value) error
```

(we can use a well-known error for "value was already set", so provider devs can ignore it if they want)

```go
// Remove removes the given value from the Set, throwing an error if it wasn't present
func (s Set) Remove(ctx context.Context, el attr.Value) error
```

?

Reply:
We can track decision making in https://github.com/hashicorp/terraform-plugin-framework/pull/126#discussion_r707360807, which will influence the answer here.